src/time/Waveform.cpp

```diff
@@ -0,0 +1,154 @@
+#include "time/Waveform.hpp"
+#include <algorithm>
+#include "cplscheme/CouplingScheme.hpp"
+#include "logging/LogMacros.hpp"
+#include "time/Time.hpp"
+#include "utils/EigenHelperFunctions.hpp"
+
+namespace precice {
+namespace time {
+
+Waveform::Waveform(
+    const int interpolationOrder)
+    : _interpolationOrder(interpolationOrder)
+{
+  PRECICE_ASSERT(not _storageIsInitialized);
+  PRECICE_ASSERT(_interpolationOrder == Time::UNDEFINED_INTERPOLATION_ORDER || (0 <= _interpolationOrder && _interpolationOrder <= 2));
```

Comment:
Also WIP and something we need to discuss. I think it makes sense to allow the user to explicitly ask for interpolation order 0, 1 etc. At the same time we have to figure out a reasonable default behavior. I think currently order 0 is a reasonable default ( = non-breaking ). In the future order 1 might be a better default (comes without additional cost and should perform better).

With this rationale it might be more reasonable to change `UNDEFINED_INTERPOLATION_ORDER` to `DEFAULT_INTERPOLATION_ORDER` and pick the `DEFAULT_INTEPOLATION_ORDER`, if nothing is defined during configuration. This allows us to easily change the default later. I think we learned that you always interpolate (implicitly with order zero, if you don't care).

Reply:
> I think it makes sense to allow the user to explicitly ask for interpolation order 0, 1 etc. At the same time we have to figure out a reasonable default behavior.

Agree

I think it might be easier if we handle the default directly in the configuration. Simply as default value of the attribute `waveform-order`.

Reply:
I cleaned it up in 0db653b73155568e2531885794f977d948832cdf. I would still like to keep the default value in the constructor of `ReadDataContext`, because then I don't have to update the tests and this will also help us in the future, when we move to a different default value.